lib/spack/spack/platforms/cray.py

```diff
@@ -119,22 +119,64 @@ def _default_target_from_env(self):
         if getattr(self, 'default', None) is None:
             bash = Executable('/bin/bash')
             output = bash(
-                '-lc', 'echo $CRAY_CPU_TARGET',
+                '--norc', '--noprofile', '-lc', 'echo $CRAY_CPU_TARGET',
                 env={'TERM': os.environ.get('TERM', '')},
-                output=str,
-                error=os.devnull
+                output=str, error=os.devnull
             )
-            output = ''.join(output.split())  # remove all whitespace
-            if output:
-                self.default = output
-                tty.debug("Found default module:%s" % self.default)
-        return self.default
+            default_from_module = ''.join(output.split())  # rm all whitespace
+            if default_from_module:
+                tty.debug("Found default module:%s" % default_from_module)
+                return default_from_module
+            else:
+                front_end = cpu.host().name
+                if front_end in list(
+                        map(lambda x: _target_name_from_craype_target_name(x),
+                            self._avail_targets())
+                ):
+                    tty.debug("default to front-end architecture")
+                    return cpu.host().name
+                else:
+                    return platform.machine()
 
     def _avail_targets(self):
         '''Return a list of available CrayPE CPU targets.'''
+
+        def modules_in_output(output):
+            """Returns a list of valid modules parsed from modulecmd output"""
+            return [i for i in re.split(r'\s\s+|\n', output)]
+
+        def target_names_from_modules(modules):
+            # Craype- module prefixes that are not valid CPU targets.
```

Comment:
Is it better to blacklist or whitelist this?  can we do it instead by looking at which modules start with `craype-<CPU family>`?  This seems a bit nasty and potentially brittle

Reply:
Re-worked completely